keps/sig-network/0029-nodelocal-dns-cache.md

```diff
@@ -0,0 +1,141 @@
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+* [Drawbacks [optional]](#drawbacks-optional)
+* [Alternatives [optional]](#alternatives-optional)
+
+[Tools for generating]: https://github.com/ekalinin/github-markdown-toc
+
+## Summary
+
+This proposal aims to improve DNS performance by running a dns caching agent on cluster nodes as a Daemonset. In today's architecture, pods in ClusterFirst DNS mode reach out to a kube-dns serviceIP for DNS queries. This is translated to a kube-dns endpoint via iptables rules added by kube-proxy. With this new architecture, pods will reach out to the dns caching agent running on the same node, thereby avoiding iptables DNAT rules and connection tracking. The local caching agent will query kube-dns for cache misses of cluster.local hostnames.
+
+
+## Motivation
+
+* With the current DNS achitecture, it is possible that pods with the highest DNS QPS have to reach out to a different node, if there is no local kube-dns instance.  
+Having a local cache will help improve the latency in such scenarios. 
+
+* Skipping iptables DNAT and connection tracking will help reduce [conntrack races](https://github.com/kubernetes/kubernetes/issues/56903) and avoid UDP DNS entries filling up conntrack table.
+
+* Connections from local caching agent to kube-dns can be upgraded to TCP. TCP conntrack entries will be removed on connection close in contrast with UDP entries that have to timeout ([default](https://www.kernel.org/doc/Documentation/networking/nf_conntrack-sysctl.txt) `nf_conntrack_udp_timeout` is 30 seconds)
+
+* Metrics & visibility into dns requests at a node level.
+
+* Neg caching can be re-enabled, thereby reducing number of queries to kube-dns.
+
+* There are several open github issues proposing a local DNS Cache daemonset and scripts to run it:
+	* [https://github.com/kubernetes/kubernetes/issues/7470#issuecomment-248912603](https://github.com/kubernetes/kubernetes/issues/7470#issuecomment-248912603)
+
+	* [https://github.com/kubernetes/kubernetes/issues/32749](https://github.com/kubernetes/kubernetes/issues/32749)
+
+	* [https://github.com/kubernetes/kubernetes/issues/45363](https://github.com/kubernetes/kubernetes/issues/45363)
+
+
+This shows that there is interest in the wider Kubernetes community for a solution similar to the proposal here. 
+
+
+### Goals
+
+Being able to run a dns caching agent as a Daemonset and get pods to use the local instance. Having visibility into cache stats and other metrics.
+
+### Non-Goals
+
+Providing a replacement for kube-dns.
+
+## Proposal
+
+A nodeLocal dns cache runs on all cluster nodes. This is managed as an add-on, runs as a Daemonset. All pods using clusterDNS will now talk to the nodeLocal cache, which will query kube-dns in case of cache misses in cluster.local domain. The node's resolv.conf will be used by this local agent for all other cache misses. 
```

Comment:
Any specific reason to use the node's resolv.conf for all misses instead of using kube-dns for all query misses?

Reply:
Wouldn't that cause problems with this?

https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#pod-s-dns-config

Reply:
The reason to use node's resolv.conf is to make use of any additional caching that the node's upstream servers might have. Since kubeDNS pods use node's resolv.conf for looking up hostnames outside cluster.local, querying those servers directly reduces queries to kube-dns, so fewer DNAT. Unless nodes running kube dns have different /etc/resolv.conf contents than other nodes.

This shouldn't affect the custom dnsConfig since only pods in clusterFirst dnsPolicy mode will get nodelocal cache ip as their dns server. custom dnsConfig pods have a dnsPolicy of "None"

Reply:
Does this mean we need to push all the stub-domain configs down to the per-node caches?

Reply:
Yes, stubDomains and custom upstream servers will need to be passed down to the per-node caches.